### PR TITLE
OSD-10251 onboard KubePersistentVolumeFillingUp to OAO in stg

### PR DESCRIPTION
**Pull request by anjoshi24**
PR changes : 

1. A new sub-directory in MCC: sre-prometheus to hold a PrometheusRule having send_managed_notification=true for ocm-agent along with a corresponding config.yaml 
2. A new sub-directory in MCC to hold the ManagedNotifications CR and a corresponding config.yaml file for staging/integration environments.

### CODE REVIEW

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+  name: sre-ocm-agent-alerts
+  namespace: openshift-monitoring
+spec:
+  groups:
+  - name: sre-ocm-agent-alerts
+    rules:
+    - alert: KubePersistentVolumeFillingUp
+    #
+      expr: (kubelet_volume_stats_available_bytes{job="kubelet",metrics_path="/metrics"} / kubelet_volume_stats_capacity_bytes{job="kubelet",metrics_path="/metrics"}) < 0.03 and kubelet_volume_stats_used_bytes{job="kubelet",metrics_path="/metrics"} > 0 unless on(namespace, persistentvolumeclaim) kube_persistentvolumeclaim_access_mode{access_mode="ReadOnlyMany"} == 1 unless on(namespace, persistentvolumeclaim) kube_persistentvolumeclaim_labels{label_alerts_k8s_io_kube_persistent_volume_filling_up="disabled"} == 1
+      for: 1m
```
Comment (mrbarge) [inline deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml:16]: This does not need to be the same content as the `KubePersistentVolumeFillingUp`. It needs to be an alert that fires when `KubePersistentVolumeFillingUp` fires for the `openshift-logging` namespace.

Something like (please test this..)

```suggestion
      expr: ALERTS{alertname="KubePersistentVolumeFillingUp", alertstate="firing", namespace="openshift-logging"} == 1
      for: 30m
```

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  labels:
+    prometheus: sre-ocm-agent-alerts
+    role: alert-rules 
+  name: sre-ocm-agent-alerts
+  namespace: openshift-monitoring
+spec:
+  groups:
+  - name: sre-ocm-agent-alerts
+    rules:
+    - alert: KubePersistentVolumeFillingUp
```
Comment (mrbarge) [inline deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml:13]: Maybe something like 

```suggestion
    - alert: LoggingVolumeFillingUpNotificationSRE
```

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+kind: PrometheusRule
+metadata:
+  labels:
+    prometheus: sre-ocm-agent-alerts
+    role: alert-rules 
+  name: sre-ocm-agent-alerts
+  namespace: openshift-monitoring
+spec:
+  groups:
+  - name: sre-ocm-agent-alerts
```
Comment (mrbarge) [inline deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml:11]:
```suggestion
  - name: sre-managed-notification-alerts
```

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
+apiVersion: monitoring.coreos.com/v1
+kind: PrometheusRule
+metadata:
+  labels:
+    prometheus: sre-ocm-agent-alerts
```
Comment (mrbarge) [inline deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml:5]:
```suggestion
    prometheus: sre-managed-notification-alerts
```

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+    role: alert-rules 
+  name: sre-ocm-agent-alerts
```
Comment (mrbarge) [inline deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml:7]:
```suggestion
  name: sre-sre-managed-notification-alerts
```

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+apiVersion: ocmagent.managed.openshift.io/v1alpha1
+kind: ManagedNotifications
+metadata:
+  name: managednotifications
+  namespace: openshift-ocm-agent-operator
+spec:
+  notification:
+    - activeBody: |-
+      This is a notification when an alert is firing.
```
Comment (mrbarge) [inline deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml:9]: Will need to change obviously but can be left until other things are addressed.

**Review by mrbarge — COMMENTED** [review]
Some feedback from a quick initial pass over things.

The `ManagedNotification` YAML is invalid and that's why the `make` is failing - I recommend before updating the PR please verify that both the PrometheusRule and the ManagedNotification are causing `ocm-agent` to behave as you expect on your own personal cluster.

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml (hunk 6 above)
Comment (mrbarge) [reply deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml:9]:
```suggestion
      Your cluster requires you to take action as its ElasticSearch cluster logging deployment has been detected as reaching a high disk usage threshold. Red Hat SRE strongly recommends reducing application logging on your cluster to ensure logging continues to function. If logging disk consumption exceeds 95%, data will be at risk of becoming unavailable or lost and the stability of your ElasticSearch deployment may be impacted.
```

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+    name: ExampleNotification
+    resendWait: 24
+    resolvedBody: |-
+      This is a notification when an alert is resolved.
```
Comment (mrbarge) [inline deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml:13]:
```suggestion
      Your cluster's ElasticSearch deployment is detected as being at safe disk consumption levels and no additional action on this issue is required.
```

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+    severity: Info
+    summary: 'Action Required: Notification'
```
Comment (mrbarge) [inline deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml:15]:
```suggestion
    summary: "ElasticSearch reaching disk capacity"
```

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+apiVersion: ocmagent.managed.openshift.io/v1alpha1
+kind: ManagedNotifications
+metadata:
+  name: managednotifications
```
[the remark this answers is not in the document]
Comment (anjoshi24) [reply deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml]: Done

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (anjoshi24) [reply deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml]: Fixed

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml [hunk 10]
```diff
@@ -0,0 +1,15 @@
+apiVersion: ocmagent.managed.openshift.io/v1alpha1
+kind: ManagedNotifications
```
[the remark this answers is not in the document]
Comment (anjoshi24) [reply deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml]: Fixed

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+metadata:
+  name: managednotifications
+  namespace: openshift-ocm-agent-operator
+spec:
+  notification:
+    - activeBody: |-
+      This is a notification when an alert is firing.
+    name: ExampleNotification
```
[the remark this answers is not in the document]
Comment (anjoshi24) [reply deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml]: Fixed

File: deploy/ocm-agent-operator-managedotifications-config/config.yaml [hunk 12]
```diff
@@ -0,0 +1,11 @@
...
+selectorSyncSet:
+  resourceApplyMode: Sync
+  matchExpressions:
+  - key: api.openshift.com/environment
+    operator: In
+    values: ["staging", "integration"]
+  #- key: api.openshift.com/fedramp
+  #  operator: NotIn
+  #  values:
+  #    - "true"
```
[the remark this answers is not in the document]
Comment (anjoshi24) [reply deploy/ocm-agent-operator-managedotifications-config/config.yaml]: Done

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+  - name: sre-ocm-agent-alerts
+    rules:
+    - alert: KubePersistentVolumeFillingUp
+    #
+      expr: (kubelet_volume_stats_available_bytes{job="kubelet",metrics_path="/metrics"} / kubelet_volume_stats_capacity_bytes{job="kubelet",metrics_path="/metrics"}) < 0.03 and kubelet_volume_stats_used_bytes{job="kubelet",metrics_path="/metrics"} > 0 unless on(namespace, persistentvolumeclaim) kube_persistentvolumeclaim_access_mode{access_mode="ReadOnlyMany"} == 1 unless on(namespace, persistentvolumeclaim) kube_persistentvolumeclaim_labels{label_alerts_k8s_io_kube_persistent_volume_filling_up="disabled"} == 1
+      for: 1m
+      labels:
+        severity: critical
+        namespace: openshift-logging
+        send_managed_notification: "true"
```
[the remark this answers is not in the document]
Comment (anjoshi24) [reply deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml]: Done

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+spec:
+  groups:
+  - name: sre-ocm-agent-alerts
+    rules:
+    - alert: KubePersistentVolumeFillingUp
+    #
+      expr: (kubelet_volume_stats_available_bytes{job="kubelet",metrics_path="/metrics"} / kubelet_volume_stats_capacity_bytes{job="kubelet",metrics_path="/metrics"}) < 0.03 and kubelet_volume_stats_used_bytes{job="kubelet",metrics_path="/metrics"} > 0 unless on(namespace, persistentvolumeclaim) kube_persistentvolumeclaim_access_mode{access_mode="ReadOnlyMany"} == 1 unless on(namespace, persistentvolumeclaim) kube_persistentvolumeclaim_labels{label_alerts_k8s_io_kube_persistent_volume_filling_up="disabled"} == 1
+      for: 1m
+      labels:
+        severity: critical
```
[the remark this answers is not in the document]
Comment (anjoshi24) [reply deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml]: Fixed

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml (hunk 1 above)
Comment (anjoshi24) [reply deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml]: Fixed

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml (hunk 3 above)
Comment (anjoshi24) [reply deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml]: Done

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml (hunk 4 above)
Comment (anjoshi24) [reply deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml]: Done

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml (hunk 5 above)
Comment (anjoshi24) [reply deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml]: Done

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml (hunk 6 above)
Comment (anjoshi24) [reply deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml]: Fixed

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml (hunk 7 above)
Comment (anjoshi24) [reply deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml]: Fixed

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml (hunk 8 above)
Comment (anjoshi24) [reply deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml]: Fixed

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml (hunk 2 above)
Comment (anjoshi24) [reply deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml]: Fixed

File: deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml [hunk 15]
```diff
@@ -0,0 +1,15 @@
+apiVersion: ocmagent.managed.openshift.io/v1alpha1
+kind: ManagedNotification
+metadata:
+  name: sre-managed-notifications
+  namespace: openshift-ocm-agent-operator
+spec:
+  notification:
```
Comment (a7vicky) [inline deploy/ocm-agent-operator-managedotifications-config/10-managednotifications-cr.yaml:7]:
```suggestion
  notifications:
```

**Comment by a7vicky** [comment]
I have tested and it's working well.
/lgtm

File: deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+  name: sre-managed-notification-alerts
+  namespace: openshift-monitoring
+spec:
+  groups:
+  - name: sre-managed-notification-alerts
+    rules:
+    - alert: LoggingVolumeFillingUpNotificationSRE
+    # KubePersistentVolumeFillingUp alert firing in openshift-logging Namespace.
+      expr: ALERTS{alertname="KubePersistentVolumeFillingUp", alertstate="firing", namespace="openshift-logging"} == 1
+      for: 1m
```
Comment (mrbarge) [inline deploy/sre-prometheus/ocm-agent/100-ocm-agent.PrometheusRule.yaml:16]: As a safe-guard against flapping alerts I think we should not trigger a notification until we know that the alert has been firing consistently for maybe 30m. (arbitrary number, but I feel it should at least be more than 1m)

```suggestion
      for: 30m
```

**Comment by mrbarge** [comment]
Two small things:

- Typo in directory name: `deploy/ocm-agent-operator-managedotifications-config`  (`otifications`)
- Can you add a `README.md` in that same directory which briefly describes what its purpose is, and who the point of contact should be (`@ocm-agent-operator` / Team Hulk). And can you also do the same for `sre-prometheus/ocm-agent/README.md`

**Pull request merged**